### PR TITLE
8343430: RISC-V: C2: Remove old trampoline call

### PR DESCRIPTION
**Pull request by RealFYang**
Hi, please consider this cleanup change.

The old trampoline call on linux-riscv64 is disabled by default after: https://bugs.openjdk.org/browse/JDK-8332689.
And the new solution (load target address from stub and call) has been baked in jdk master for quite some time.
This propose to remove the old trampoline call related logic so that the code will be cleaner and easier to maintain.
As we are still using `trampoline_stub_Relocation` for the stub, I kept using the original `emit_trampoline_stub` name.

Testing on linux-riscv64:
- [x] tier1-tier3 (release)
- [x] hotspot:tier1 (fastdebug)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8343430](https://bugs.openjdk.org/browse/JDK-8343430): RISC-V: C2: Remove old trampoline call (**Enhancement** - P4)

### Reviewers
 * [Robbin Ehn](https://openjdk.org/census#rehn) (@robehn - **Reviewer**) Review applies to [ad6df2fe](https://git.openjdk.org/jdk/pull/21822/files/ad6df2fea1654eb53ea8ec243cb8ccd1f6a2a00b)
 * [Hamlin Li](https://openjdk.org/census#mli) (@Hamlin-Li - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21822/head:pull/21822` \
`$ git checkout pull/21822`

Update a local copy of the PR: \
`$ git checkout pull/21822` \
`$ git pull https://git.openjdk.org/jdk.git pull/21822/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21822`

View PR using the GUI difftool: \
`$ git pr show -t 21822`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21822.diff">https://git.openjdk.org/jdk/pull/21822.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21822#issuecomment-2453745516)
</details>

### CODE REVIEW

**Comment by robehn** [comment]
Can we change the name to something like address or far address stub?

**Comment by RealFYang** [comment]
> Can we change the name to something like address or far address stub?
> 
> I think this comment, `Emit a trampoline stub for a call to a target` would be much easier to understand what really happens, if we say address stub.

Hmm... What about names like `max_reloc_call_stub_size` and `reloc_call_address_stub`?

**Comment by robehn** [comment]
Yea, reloc_call_address_stub, sounds nice.

File: src/hotspot/cpu/riscv/c1_LIRAssembler_riscv.hpp [hunk 1]
```diff
@@ -65,10 +65,9 @@ friend class ArrayCopyStub;
   void deoptimize_trap(CodeEmitInfo *info);
 
   enum {
-    // See emit_static_call_stub for detail
-    // CompiledDirectCall::to_interp_stub_size() (14) + CompiledDirectCall::to_trampoline_stub_size() (1 + 3 + address)
-    _call_stub_size = 14 * MacroAssembler::instruction_size +
-                      (MacroAssembler::instruction_size + MacroAssembler::NativeShortCall::trampoline_size),
+    // call stub: CompiledDirectCall::to_interp_stub_size() +
+    //            CompiledDirectCall::to_trampoline_stub_size()
+    _call_stub_size = 11 * MacroAssembler::instruction_size + MacroAssembler::NativeFarCall::trampoline_size,
```
Comment (robehn) [inline src/hotspot/cpu/riscv/c1_LIRAssembler_riscv.hpp:70]: I think this should call_address_stub also?

File: src/hotspot/cpu/riscv/compiledIC_riscv.cpp [hunk 2]
```diff
@@ -69,9 +69,9 @@ int CompiledDirectCall::to_interp_stub_size() {
 }
 
 int CompiledDirectCall::to_trampoline_stub_size() {
-  // We count instructions and an additional alignment nop.
+  // We count size of target address and an additional alignment nop.
   // Trampoline stubs are always word aligned.
```
Comment (robehn) [inline src/hotspot/cpu/riscv/compiledIC_riscv.cpp:73]: Call address stub?

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp [hunk 3]
```diff
@@ -1636,9 +1593,8 @@ class MacroAssembler: public Assembler {
     load_pc_relative_instruction_size = 2 * instruction_size // auipc, ld
   };
 
-  enum NativeShortCall {
-    trampoline_size        = 3 * instruction_size + wordSize,
-    trampoline_data_offset = 3 * instruction_size
+  enum NativeFarCall {
+    trampoline_size        = 1 * instruction_size + wordSize,
```
Comment (robehn) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.hpp:1597]: Call address stub size?

**Review by robehn — COMMENTED** [review]
Thanks! I think there are some more places to fix naming.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp (hunk 3 above)
Comment (RealFYang) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.hpp]: I am a bit worried that `MacroAssembler::NativeShortCall` could be confusing with `MacroAssembler::far_call` in naming, so I simply removed this enum. I also improved the code comment about different types of calls. Let me know.

**Review by robehn — APPROVED** [review]
Thanks, I think it looks good.

**Comment by RealFYang** [comment]
> Thanks, I think it looks good.

Thanks for the suggestions. I will wait for another reviewer. @Hamlin-Li : Do you want to take a look?

**Review by Hamlin-Li — APPROVED** [review]
Looks good. Thanks.

**Comment by RealFYang** [comment]
Hi, I just pushed a small change addressing two remaining minor issues which I think should be resolved in this PR.

1. The distance calculation in MacroAssembler::reloc_call is confusing and is not necessary;
   We are not doing a pc-relative loading from the target, but from the stub.

2. The reloc_call in riscv_enc_java_to_runtime could be further simplified into a simple far_call.
   This reloc_call is there for the old trampoline call. Also updated related code comment.

@robehn @Hamlin-Li  : I hope that still looks good to you. Thanks.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 4]
```diff
@@ -4278,8 +4219,12 @@ address MacroAssembler::load_and_call(Address entry) {
     assert_alignment(call_pc);
   }
 #endif
+
   relocate(entry.rspec(), [&] {
-    load_link_jump(target, t1);
+    int64_t distance = 0;
+    auipc(tmp, (int32_t)distance + 0x800);
+    ld(tmp, Address(tmp, ((int32_t)distance << 20) >> 20));
+    jalr(tmp);
```
Comment (Hamlin-Li) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:4227]: Seems we can totally remove the usage of `distance`? could be something like below:
```
auipc(tmp, 0);
ld(tmp, Address(tmp, 0));
jalr(tmp);
```

Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:4227]: If the above simplication make sense, then what we're doing here is to load from pc as the destination addr, then jump to the addr? I could be misunderstanding.

**Review by Hamlin-Li — COMMENTED** [review]
Thanks for updating. Sorry for the delay.
Just had another look, got some questions.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp (hunk 4 above)
Comment (RealFYang) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp]: Thanks. I removed `distance` and simplified the code as suggested. Note that the use of PC as address here is only temporary. The relocation created while emitting the stub will ensure this call sequence is subsequently patched to call the stub. I have added code comments for to help understand.

Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp]: Thanks for updating and explanation.

**Review by Hamlin-Li — APPROVED** [review]
Updating also looks good to me. Thanks!

**Review by robehn — APPROVED** [review]

**Comment by RealFYang** [comment]
Thanks. I will wait a while for JEP 491 to land as I see this has conflict with that big change.

**Comment by RealFYang** [comment]
@Hamlin-Li @robehn : Hello, I just merged with master resolving a small conflict in code comment. I think we are still fine. Could you please re-approve this? Thanks.

**Review by Hamlin-Li — APPROVED** [review]
Still good, Thanks!

**Pull request closed without merging**